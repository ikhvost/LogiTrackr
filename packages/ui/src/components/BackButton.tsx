import { useNavigate } from 'react-router-dom'

export const BackButton = () => {
  const navigate = useNavigate()

  return (
    <button
      onClick={() => navigate('/')}
      className="flex items-center px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600 transition-colors"
    >
      <svg className="h-5 w-5 mr-2" fill="none" viewBox="0 0 24 24" stroke="currentColor">
        <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M10 19l-7-7m0 0l7-7m-7 7h18" />
      </svg>
      Back
    </button>
  )
}
