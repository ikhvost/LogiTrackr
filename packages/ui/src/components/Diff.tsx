import React from 'react'
import { diffJson, Change } from 'diff'

interface Props {
  oldData: unknown
  currentData: unknown
}

enum DiffType {
  Added = 'added',
  Removed = 'removed',
  Unchanged = 'unchanged',
}

export const Diff = ({ oldData, currentData }: Props) => {
  const differences = diffJson(oldData, currentData)

  const renderLine = (line: string, type: DiffType) => {
    const added = 'bg-green-100 text-green-800'
    const removed = 'bg-red-100 text-red-800'
    const unchanged = ''

    return (
      <div className={type === DiffType.Added ? added : type === DiffType.Removed ? removed : unchanged}>{line}</div>
    )
  }

  const renderPart = (part: Change) => {
    return part.value
      .split('\n')
      .map((line, index) => (
        <React.Fragment key={index}>
          {renderLine(line, part.added ? DiffType.Added : part.removed ? DiffType.Removed : DiffType.Unchanged)}
        </React.Fragment>
      ))
  }

  return (
    <div className="flex">
      <div className="w-1/2 pr-2">
        <h4 className="text-md font-semibold mb-1">Previous Version</h4>
        <pre className="whitespace-pre-wrap bg-white p-2 rounded text-sm">
          {differences.map((part, index) => (
            <React.Fragment key={index}>{!part.added && renderPart(part)}</React.Fragment>
          ))}
        </pre>
      </div>
      <div className="w-1/2 pl-2">
        <h4 className="text-md font-semibold mb-1">Current Version</h4>
        <pre className="whitespace-pre-wrap bg-white p-2 rounded text-sm">
          {differences.map((part, index) => (
            <React.Fragment key={index}>{!part.removed && renderPart(part)}</React.Fragment>
          ))}
        </pre>
      </div>
    </div>
  )
}
