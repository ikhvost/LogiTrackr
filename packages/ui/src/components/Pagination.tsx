import { Metadata } from '@saas-versioning/contracts'

interface Props {
  metadata: Metadata
  itemsPerPage: number
  onPrevPage: () => void
  onNextPage: () => void
}

export const Pagination = ({ metadata, onPrevPage, onNextPage, itemsPerPage }: Props) => {
  return (
    <div className="flex justify-between items-center mt-8">
      <p className="text-sm text-gray-500">
        Showing {Math.min(metadata.totalCount, itemsPerPage)} of {metadata.totalCount} entries
      </p>
      <div className="flex space-x-2 items-center">
        <button
          onClick={onPrevPage}
          disabled={metadata.currentPage === 1}
          className="text-gray-600 px-4 py-2 rounded-lg bg-white hover:bg-gray-50 disabled:opacity-50 transition-all shadow-sm"
        >
          «
        </button>
        <div className="text-sm text-gray-600">
          Page {metadata.currentPage} of {metadata.totalPages}
        </div>
        <button
          onClick={onNextPage}
          disabled={metadata.currentPage === metadata.totalPages}
          className="text-gray-600 px-4 py-2 rounded-lg bg-white hover:bg-gray-50 disabled:opacity-50 transition-all shadow-sm"
        >
          »
        </button>
      </div>
    </div>
  )
}
