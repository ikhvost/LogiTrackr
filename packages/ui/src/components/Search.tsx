import { useState } from 'react'

interface Props {
  onSearch: (query: string) => void
}

export const Search = ({ onSearch }: Props) => {
  const [query, setQuery] = useState('')
  const handleSearch = () => onSearch(query)

  return (
    <div className="flex flex-col md:flex-row justify-between items-center space-y-4 md:space-y-0 md:space-x-4">
      <input
        type="text"
        placeholder="Search by ID or type..."
        className="w-full md:w-3/4 p-4 rounded-full border-0 focus:ring-2 focus:ring-indigo-300 transition text-gray-700 text-lg"
        value={query}
        onChange={(e) => setQuery(e.target.value)}
        onKeyUp={(e) => e.key === 'Enter' && handleSearch()}
      />
      <button
        onClick={handleSearch}
        className="w-full md:w-auto px-10 py-4 bg-indigo-600 text-white text-lg font-semibold rounded-full shadow-lg hover:bg-indigo-700 transition-all duration-300 ease-in-out"
      >
        Search
      </button>
    </div>
  )
}
