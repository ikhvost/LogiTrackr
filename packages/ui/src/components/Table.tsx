import { useState, Fragment, ReactNode } from 'react'

interface Column<T> {
  header: string
  accessor: keyof T
  render?: (item: T) => ReactNode
}

interface Props<T> {
  data: T[]
  columns: readonly Column<T>[]
  onRowClick?: (item: T) => void
  renderExpandedRow?: (item: T) => ReactNode
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const Table = <T extends Record<string, any>>({ data, columns, onRowClick, renderExpandedRow }: Props<T>) => {
  const [expandedRow, setExpandedRow] = useState<number | null>(null)

  const handleRowClick = (item: T, index: number) => {
    if (renderExpandedRow) {
      setExpandedRow(expandedRow === index ? null : index)
    } else if (onRowClick) {
      onRowClick(item)
    }
  }

  return (
    <table className="min-w-full divide-y divide-gray-200">
      <thead className="bg-gray-50">
        <tr>
          {columns.map((column) => (
            <th
              key={column.header}
              scope="col"
              className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
            >
              {column.header}
            </th>
          ))}
        </tr>
      </thead>
      <tbody className="bg-white divide-y divide-gray-200">
        {data.length === 0 ? (
          <tr>
            <td colSpan={columns.length} className="px-6 py-4 text-center text-gray-500">
              No records found
            </td>
          </tr>
        ) : (
          data.map((item, index) => (
            <Fragment key={index}>
              <tr
                onClick={() => handleRowClick(item, index)}
                className={`cursor-pointer hover:bg-gray-100 ${expandedRow === index ? 'bg-gray-50' : ''}`}
              >
                {columns.map((column) => (
                  <td key={column.header} className="px-6 py-4 whitespace-nowrap">
                    {column.render ? column.render(item) : String(item[column.accessor])}
                  </td>
                ))}
              </tr>
              {expandedRow === index && renderExpandedRow && (
                <tr>
                  <td colSpan={columns.length} className="px-6 py-4">
                    {renderExpandedRow(item)}
                  </td>
                </tr>
              )}
            </Fragment>
          ))
        )}
      </tbody>
    </table>
  )
}
