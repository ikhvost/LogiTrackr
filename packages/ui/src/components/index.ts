export * from './Table'
export * from './Pagination'
export * from './Search'
export * from './Diff'
export * from './BackButton'
