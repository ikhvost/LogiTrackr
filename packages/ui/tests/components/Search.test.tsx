import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { describe, test, expect, vi } from 'vitest'
import { Search } from '../../src/components/Search'

describe('<Search />', () => {
  test('rendering: component renders without crashing', () => {
    const onSearch = vi.fn()
    const { container } = render(<Search onSearch={onSearch} />)
    expect(container).toMatchSnapshot()
  })

  test('input field renders correctly', () => {
    const onSearch = vi.fn()
    render(<Search onSearch={onSearch} />)

    const input = screen.getByPlaceholderText('Search by ID or type...')
    expect(input).toBeInTheDocument()
    expect(input).toHaveClass(
      'w-full md:w-3/4 p-4 rounded-full border-0 focus:ring-2 focus:ring-indigo-300 transition text-gray-700 text-lg',
    )
  })

  test('search button renders correctly', () => {
    const onSearch = vi.fn()
    render(<Search onSearch={onSearch} />)

    const button = screen.getByText('Search')
    expect(button).toBeInTheDocument()
    expect(button).toHaveClass(
      'w-full md:w-auto px-10 py-4 bg-indigo-600 text-white text-lg font-semibold rounded-full shadow-lg hover:bg-indigo-700 transition-all duration-300 ease-in-out',
    )
  })

  test('input updates on change', () => {
    const onSearch = vi.fn()
    render(<Search onSearch={onSearch} />)

    const input = screen.getByPlaceholderText('Search by ID or type...')
    fireEvent.change(input, { target: { value: 'test query' } })
    expect(input).toHaveValue('test query')
  })

  test('onSearch is called when search button is clicked', () => {
    const onSearch = vi.fn()
    render(<Search onSearch={onSearch} />)

    const input = screen.getByPlaceholderText('Search by ID or type...')
    fireEvent.change(input, { target: { value: 'test query' } })

    const button = screen.getByText('Search')
    fireEvent.click(button)

    expect(onSearch).toHaveBeenCalledWith('test query')
  })

  test('onSearch is called when Enter key is pressed', () => {
    const onSearch = vi.fn()
    render(<Search onSearch={onSearch} />)

    const input = screen.getByPlaceholderText('Search by ID or type...')
    fireEvent.change(input, { target: { value: 'test query' } })
    fireEvent.keyUp(input, { key: 'Enter', code: 'Enter' })

    expect(onSearch).toHaveBeenCalledWith('test query')
  })

  test('onSearch is not called for other key presses', () => {
    const onSearch = vi.fn()
    render(<Search onSearch={onSearch} />)

    const input = screen.getByPlaceholderText('Search by ID or type...')
    fireEvent.change(input, { target: { value: 'test query' } })
    fireEvent.keyUp(input, { key: 'A', code: 'KeyA' })

    expect(onSearch).not.toHaveBeenCalled()
  })
})
