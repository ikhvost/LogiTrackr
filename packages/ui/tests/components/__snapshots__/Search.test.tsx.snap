// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Search /> > rendering: component renders without crashing 1`] = `
<div>
  <div
    class="flex flex-col md:flex-row justify-between items-center space-y-4 md:space-y-0 md:space-x-4"
  >
    <input
      class="w-full md:w-3/4 p-4 rounded-full border-0 focus:ring-2 focus:ring-indigo-300 transition text-gray-700 text-lg"
      placeholder="Search by ID or type..."
      type="text"
      value=""
    />
    <button
      class="w-full md:w-auto px-10 py-4 bg-indigo-600 text-white text-lg font-semibold rounded-full shadow-lg hover:bg-indigo-700 transition-all duration-300 ease-in-out"
    >
      Search
    </button>
  </div>
</div>
`;
